delfin/task_manager/metrics_manager.py

```diff
@@ -34,27 +46,288 @@ class MetricsTaskManager(manager.Manager):
+                                      self.on_process_join,
+                                      self.on_process_leave)
+        self.group = group
+        self.watcher = watcher
+        self.watch_job_id = uuidutils.generate_uuid()
+        self.scheduler.add_job(watcher.watch_group_change, 'interval',
+                               seconds=CONF.telemetry.
+                               group_change_detect_interval,
+                               next_run_time=datetime.datetime.now(),
+                               id=self.watch_job_id)
+        LOG.info('Created watch for group membership change for group {0}.'
+                 .format(group))
+        self.cleanup_job_id = uuidutils.generate_uuid()
+        self.scheduler.add_job(self.process_cleanup, 'interval',
+                               seconds=CONF.telemetry.process_cleanup_interval,
+                               next_run_time=datetime.datetime.now(),
+                               id=self.cleanup_job_id)
+        LOG.info('Created process cleanup background job for group {0}.'
+                 .format(group))
+
+    def on_process_join(self, event):
+        LOG.info('Member %s joined the group %s' % (event.member_id,
+                                                    event.group_id))
+        host = event.group_id.decode('utf-8')
+        if self.watcher:
+            LOG.info('Processes in current node {0}'
+                     .format(self.watcher.get_members(host)))
+
+    def on_process_leave(self, event):
+        LOG.info('Member %s left the group %s' % (event.member_id,
+                                                  event.group_id))
+        executor_topic = event.member_id.decode('utf-8')
+        name = executor_topic.split(':')[1]
+        if name in self.executor_map.keys():
+            host = event.group_id.decode('utf-8')
+            LOG.info("Re-create process {0} in {1} that is handling tasks"
+                     .format(executor_topic, host))
+            launcher = self.create_process(executor_topic, host)
+            self.executor_map[name]["launcher"] = launcher
+            context = ctxt.get_admin_context()
+            for storage_id in self.executor_map[name]["storages"]:
+                tasks, failed_tasks = self.get_all_tasks(storage_id)
+                for task in tasks:
+                    LOG.info("Re-scheduling task {0} of storage {1}"
+                             .format(task['id'], storage_id))
+                    self.rpcapi.assign_job_local(
+                        context, task['id'], executor_topic)
+
+                for f_task in failed_tasks:
+                    LOG.info("Re-scheduling failed failed task {0},"
+                             " of storage {1}"
+                             .format(f_task['id'], storage_id))
+                    self.rpcapi.assign_failed_job_local(
+                        context, f_task['id'], executor_topic)
+
+    def process_cleanup(self):
```

Comment:
Except the periodically cleanup, do we have another mechanism to cleanup it forwardly?

Reply:
When a storage is deleted and the process do not handle any other storages, we want stop the processes after a delay (for the process to handle remove storage). We send the remove storage message, wait for the message to be handled and cleanup later.